src/huggingface_hub/hub_mixin.py

```diff
@@ -0,0 +1,182 @@
+                Whether or not to delete incompletely received files. Will attempt to resume the download if such a
+                file exists.
+            proxies (:obj:`Dict[str, str], `optional`):
+                A dictionary of proxy servers to use by protocol or endpoint, e.g., :obj:`{'http': 'foo.bar:3128',
+                'http://hostname': 'foo.bar:4012'}`. The proxies are used on each request.
+            local_files_only(:obj:`bool`, `optional`, defaults to :obj:`False`):
+                Whether or not to only look at local files (i.e., do not try to download the model).
+            use_auth_token (:obj:`str` or `bool`, `optional`):
+                The token to use as HTTP bearer authorization for remote files. If :obj:`True`, will use the token
+                generated when running :obj:`transformers-cli login` (stored in :obj:`~/.huggingface`).
+            revision(:obj:`str`, `optional`, defaults to :obj:`"main"`):
+                The specific model version to use. It can be a branch name, a tag name, or a commit id, since we use a
+                git-based system for storing models and other artifacts on huggingface.co, so ``revision`` can be any
+                identifier allowed by git.
+        .. note::
+            Passing :obj:`use_auth_token=True` is required when you want to use a private model.
+        """
+
+        model_id = pretrained_model_name_or_path
+        strict = kwargs.pop("strict", True)
+        map_location = kwargs.pop("map_location", torch.device("cpu"))
+        revision = kwargs.pop("revision", None)
+        force_download = kwargs.pop("force_download", False)
+        resume_download = kwargs.pop("resume_download", False)
+        proxies = kwargs.pop("proxies", None)
+        use_auth_token = kwargs.pop("use_auth_token", None)
+        cache_dir = kwargs.pop("cache_dir", None)
+        local_files_only = kwargs.pop("local_files_only", False)
+
+        if len(model_id.split("/")) == 1:
+            name = model_id
+        else:
+            _, name = model_id.split("/")
+
+        if name in os.listdir() and CONFIG_NAME in os.listdir(name):
+            print("LOADING weights from local directory")
+            config_file = os.path.join(name, CONFIG_NAME)
+        else:
+            try:
+                config_url = hf_hub_url(model_id, filename=CONFIG_NAME, revision=revision)
+                config_file = cached_download(config_url, cache_dir=cache_dir, force_download=force_download, proxies=proxies, resume_download=resume_download, local_files_only=local_files_only, use_auth_token=use_auth_token)
+            except:
+                config_file = None
+
+        if name in os.listdir():
+            print("LOADING weights from local directory")
+            model_file = os.path.join(name, PYTORCH_WEIGHTS_NAME)
+        else:
+            model_url = hf_hub_url(model_id, filename=PYTORCH_WEIGHTS_NAME, revision=revision)
```

Comment:
In other implementations, the revision can also be obtained from a split of the model_id on `@`. I think it would make sense to also support that here, so that one may do:
```py
model = MyModel.from_pretrained("username/mymodel@branch")
```
As it is already the case in several libraries.

See [asteroid](https://github.com/asteroid-team/asteroid/blob/3c6abcfc0f5dafb8cf446488f340dedd4b46124d/asteroid/utils/hub_utils.py#L59-L61) or [pyannote](https://github.com/julien-c/pyannote-audio/blob/eb4a021db13c63d8d7f1230ec1c4462db780e480/pyannote/audio/core/model.py#L684-L686)